src/GraphQL/Types/TypeInformation.cs

```diff
@@ -0,0 +1,259 @@
+        /// </summary>
+        public Type? GraphType
+        {
+            get => _graphType;
+            set
+            {
+                if (value == null)
+                {
+                    _graphType = null;
+                    return;
+                }
+                if (IsInputType)
+                {
+                    if (!value.IsInputType())
+                    {
+                        throw new ArgumentOutOfRangeException(nameof(value), "Value can only be an input graph type.");
+                    }
+                }
+                else if (!value.IsOutputType())
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value), "Value can only be an output graph type.");
+                }
+                if (!value.IsNamedType())
+                {
+                    throw new ArgumentOutOfRangeException(nameof(value), "Value must be a named graph type.");
+                }
+                _graphType = value;
+            }
+        }
+
+        /// <summary>
+        /// Initializes an instance with the specified properties.
+        /// </summary>
+        /// <param name="memberInfo">The member being inspected.</param>
+        /// <param name="isInputType">Indicates that this is an input type (an argument); false for output types.</param>
+        /// <param name="type">The underlying type.</param>
+        /// <param name="isNullable">Indicates that the underlying type is nullable.</param>
+        /// <param name="isList">Indicates that this represents a list of elements.</param>
+        /// <param name="listIsNullable">Indicates that the list is nullable.</param>
+        /// <param name="graphType">The graph type of the underlying CLR type; null if not specified.</param>
+        public TypeInformation(MemberInfo memberInfo, bool isInputType, Type type, bool isNullable, bool isList, bool listIsNullable, Type? graphType)
+        {
+            MemberInfo = memberInfo;
+            IsInputType = isInputType;
+            Type = type;
+            IsNullable = isNullable;
+            IsList = isList;
+            ListIsNullable = listIsNullable;
+            GraphType = graphType;
+        }
+
+        /// <summary>
+        /// Initializes an instance containing type information necessary to select a graph type.
+        /// The instance is populated based on inspecting the type and NRT annotations on the specified property.
+        /// </summary>
+        public TypeInformation(PropertyInfo propertyInfo, bool isInputProperty)
+            : this(propertyInfo, isInputProperty, propertyInfo.PropertyType, false, false, false, null)
+        {
+            var typeTree = Interpret(new NullabilityInfoContext().Create(propertyInfo), isInputProperty);
+
+            foreach (var type in typeTree)
+            {
+                //detect list types, but not lists of lists
```

Comment:
How to deal with lists of lists?

Reply:
I think `int[]?[]?` should end up as `ListGraphType<GraphQLClrOutputTypeReference<int[]>>` - but probably doesn't now and I should fix that and add a test.

Reply:
GraphQL 100% supports lists of lists, I explicitly asked about it in the spec repo. OK to merge as is for now.

Reply:
Actually GraphQL allows for lists of lists.... so maybe I should leave it alone for now...

Reply:
Ok yeah, I forgot.  Ok I won't touch it, but I will add a test to show existing behavior, which does support lists of lists, but not lists of lists of lists, and the second level of lists won't have the right NRT attribute..

Reply:
Keep in mind that we can use `[InputType]` or `[OutputType]` to force a specific exact graph type, so all is not lost...